src/envs/pybullet_blocks.py

```diff
@@ -0,0 +1,728 @@
+                                   types=types,
+                                   params_space=params_space,
+                                   policy=_policy,
+                                   initiable=lambda _1, _2, _3, _4: True,
+                                   terminal=_terminal)
+
+    def _move_relative_to_block(self, name: str,
+                                rel_or_abs_target_pose: Pose3D,
+                                rel_or_abs: Tuple[str, str,
+                                                  str], types: Sequence[Type],
+                                params_space: Box) -> ParameterizedOption:
+        """Creates a ParameterizedOption for moving to a target pose relative
+        to a block.
+
+        Each of the three dimensions of the target pose can be specified
+        relatively or absolutely. The rel_or_abs argument indicates
+        whether each dimension is relative ("rel") or absolute ("abs").
+        """
+        assert types == [self._robot_type, self._block_type]
+
+        def _get_current_and_target_pose(
+                state: State, objects: Sequence[Object],
+                params: Array) -> Tuple[Pose3D, Pose3D]:
+            del params  # not used
+            robot, block = objects
+            current_pose = (state[robot][0], state[robot][1], state[robot][2])
+            block_pose = (state[block][0], state[block][1], state[block][2])
+            target_pose = self._convert_rel_abs_to_abs(block_pose,
+                                                       rel_or_abs_target_pose,
+                                                       rel_or_abs)
+            return current_pose, target_pose
+
+        return self._move_to_pose(name, types, params_space,
+                                  _get_current_and_target_pose)
+
+    def _change_grippers(self, name: str, target_val: float,
+                         types: Sequence[Type],
+                         params_space: Box) -> ParameterizedOption:
+        """Creates a ParameterizedOption for changing the robot grippers."""
+
+        assert types[0] == self._robot_type
+
+        def _policy(state: State, memory: Dict, objects: Sequence[Object],
+                    params: Array) -> Action:
+            del memory, objects, params  # unused
+            current_val = state.get(self._robot, "fingers")
+            gripper_action = target_val - current_val
+            gripper_action = np.clip(gripper_action, self.action_space.low[3],
+                                     self.action_space.high[3])
+            return Action(
+                np.array([0., 0., 0., gripper_action], dtype=np.float32))
+
+        def _terminal(state: State, memory: Dict, objects: Sequence[Object],
+                      params: Array) -> bool:
+            del memory, objects, params  # unused
+            current_val = state.get(self._robot, "fingers")
+            dist = (target_val - current_val)**2
+            return dist < self._grasp_tol
```

Comment:
it's interesting that grasp_tol has to be 0.05, while move_tol can be so tiny. any insight into why?

Reply:
also, this is again the squared dist, update name accordingly

Reply:
it's because the fingers are way more fickle than the end effector (related to the nudging stuff).